File: python/pilot_utils.py
class PilotLFSR:
    def __init__(self):
        self.INITIAL_STATE = 0x7F
        self.state = self.INITIAL_STATE
        self.POLARITY = [1, 1, 1, -1]

    def next_polarity(self):
        out_bit = (self.state >> 6) & 1
        feedback = out_bit ^ ((self.state >> 3) & 1)
        self.state = ((self.state << 1) | feedback) & 0x7F
        return feedback if out_bit == -1.0 else 1.0
    
    def reset(self):
        self.state = self.INITIAL_STATE

File: cpp/src/pilot_utils.cpp
//
// Created by Kanishka on 4/19/2026.
//

#include "phy/pilot_utils.h"

namespace wifi80211a {

    double PilotLFSR::next_polarity()
    {
        // Taps at positions 7 and 4 (1-indexed) → bits 6 and 3 in a 7-bit register.
        const int out_bit  = (state_ >> 6) & 1;
        const int feedback = out_bit ^ ((state_ >> 3) & 1);
        state_ = static_cast<uint8_t>(((state_ << 1) | feedback) & 0x7F);
        return feedback ? -1.0 : +1.0; // BPSK: 0 → +1, 1 → −1
    }

    void PilotLFSR::reset()
    {
        state_ = INITIAL_STATE;
    }

} // namespace wifi80211a
